File: app/src/main/java/com/drdisagree/colorblendr/extension/ThemeOverlayPackage.java
package com.drdisagree.colorblendr.extension;

import static com.drdisagree.colorblendr.common.Const.MONET_SEED_COLOR;
import static com.drdisagree.colorblendr.common.Const.MONET_STYLE_ORIGINAL_NAME;

import android.graphics.Color;
import android.os.Build;
import android.util.Log;

import com.drdisagree.colorblendr.config.RPrefs;
import com.drdisagree.colorblendr.utils.ColorUtil;

import org.json.JSONObject;

public class ThemeOverlayPackage {

    private static final String TAG = ThemeOverlayPackage.class.getSimpleName();
    public static final String THEME_STYLE = "android.theme.customization.theme_style";
    public static final String COLOR_SOURCE = "android.theme.customization.color_source";
    public static final String SYSTEM_PALETTE = "android.theme.customization.system_palette";
    public static final String ACCENT_COLOR = "android.theme.customization.accent_color";
    public static final String COLOR_BOTH = "android.theme.customization.color_both";
    public static final String APPLIED_TIMESTAMP = "_applied_timestamp";

    public static JSONObject getThemeCustomizationOverlayPackages() {
        JSONObject object = new JSONObject();

        try {
            object.putOpt(
                    THEME_STYLE,
                    RPrefs.getString(MONET_STYLE_ORIGINAL_NAME, "TONAL_SPOT")
            );
            object.putOpt(COLOR_SOURCE, "preset");
            object.putOpt(
                    SYSTEM_PALETTE,
                    ColorUtil.intToHexColorNoHash(RPrefs.getInt(MONET_SEED_COLOR, Color.BLUE))
            );
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
                object.putOpt(
                        ACCENT_COLOR,
                        ColorUtil.intToHexColorNoHash(RPrefs.getInt(MONET_SEED_COLOR, Color.BLUE))
                );
            }
            object.putOpt(APPLIED_TIMESTAMP, String.valueOf(System.currentTimeMillis()));
        } catch (Exception e) {
            Log.e(TAG, "getThemeCustomizationOverlayPackages:", e);
        }

        return object;
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/extension/ThemeOverlayPackage.kt
package com.drdisagree.colorblendr.extension

import android.graphics.Color
import android.os.Build
import android.util.Log
import com.drdisagree.colorblendr.data.common.Utilities.getOriginalStyleName
import com.drdisagree.colorblendr.data.common.Utilities.getSeedColorValue
import com.drdisagree.colorblendr.utils.colors.ColorUtil.intToHexColorNoHash
import org.json.JSONObject

object ThemeOverlayPackage {

    private val TAG: String = ThemeOverlayPackage::class.java.simpleName
    const val THEME_STYLE: String = "android.theme.customization.theme_style"
    const val COLOR_SOURCE: String = "android.theme.customization.color_source"
    const val SYSTEM_PALETTE: String = "android.theme.customization.system_palette"
    const val ACCENT_COLOR: String = "android.theme.customization.accent_color"
    const val COLOR_BOTH: String = "android.theme.customization.color_both"
    const val APPLIED_TIMESTAMP: String = "_applied_timestamp"

    val themeCustomizationOverlayPackages: JSONObject
        get() {
            return try {
                JSONObject().apply {
                    putOpt(
                        COLOR_SOURCE,
                        "preset"
                    )
                    putOpt(
                        THEME_STYLE,
                        getOriginalStyleName()
                    )
                    putOpt(
                        SYSTEM_PALETTE,
                        intToHexColorNoHash(getSeedColorValue(Color.BLUE))
                    )
                    if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
                        putOpt(
                            ACCENT_COLOR,
                            intToHexColorNoHash(getSeedColorValue(Color.BLUE))
                        )
                    }
                    putOpt(
                        APPLIED_TIMESTAMP,
                        System.currentTimeMillis()
                    )
                }
            } catch (e: Exception) {
                Log.e(TAG, "themeCustomizationOverlayPackages:", e)
                JSONObject()
            }
        }

    fun getOriginalSettings(currentSettings: String): JSONObject {
        return JSONObject(currentSettings).apply {
            val keysToRemove = arrayOf(
                THEME_STYLE,
                COLOR_SOURCE,
                SYSTEM_PALETTE
            )

            for (key in keysToRemove) {
                remove(key)
            }

            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.S_V2) {
                remove(ACCENT_COLOR)
            }

            putOpt(COLOR_BOTH, "0")
            putOpt(COLOR_SOURCE, "home_wallpaper")
            putOpt(APPLIED_TIMESTAMP, System.currentTimeMillis())
        }
    }
}
